src/patches.js

```diff
@@ -6,50 +6,89 @@ export function generatePatches(state, basePath, patches, inversePatches) {
         : generateObjectPatches(state, basePath, patches, inversePatches)
 }
 
-function generateArrayPatches(state, basePath, patches, inversePatches) {
-    const {base, copy, assigned} = state
-    const minLength = Math.min(base.length, copy.length)
+function _generateArrayPatches(
+    base,
+    copy,
+    assigned,
+    basePath,
+    patches,
+    inversePatches
+) {
+    if (copy.length < base.length) {
+        ;[copy, base] = [base, copy]
+        ;[patches, inversePatches] = [inversePatches, patches]
+    }
 
-    // Look for replaced indices.
-    for (let i = 0; i < minLength; i++) {
-        if (assigned[i] && base[i] !== copy[i]) {
-            const path = basePath.concat(i)
-            patches.push({op: "replace", path, value: copy[i]})
-            inversePatches.push({op: "replace", path, value: base[i]})
-        }
+    const delta = copy.length - base.length
+
+    // Find the first changed index.
+    let start = 0
+    while (base[start] === copy[start] && start < base.length) {
+        ++start
+    }
+
+    // Find the first changed index from right
+    let baseEnd = base.length
+    while (baseEnd > start && base[baseEnd - 1] === copy[baseEnd + delta - 1]) {
+        --baseEnd
     }
 
-    // Did the array expand?
-    if (minLength < copy.length) {
-        for (let i = minLength; i < copy.length; i++) {
+    // Look for replaced indices.
+    const replaceCount = baseEnd - start
+    for (let i = 0; i < replaceCount; ++i) {
+        const index = start + i
+        const path = basePath.concat([index])
+        if (assigned[index] && copy[index] !== base[index]) {
             patches.push({
-                op: "add",
-                path: basePath.concat(i),
-                value: copy[i]
+                op: "replace",
+                path,
+                value: copy[index]
+            })
+            inversePatches.push({
+                op: "replace",
+                path,
+                value: base[index]
             })
         }
-        inversePatches.push({
-            op: "replace",
-            path: basePath.concat("length"),
-            value: base.length
-        })
     }
+    start += replaceCount
 
-    // ...or did it shrink?
-    else if (minLength < base.length) {
+    // If add items to end, we can use replace 'length' property
+    // instead of multiple remove operations
```

Comment:
Use this comment:

```js
// For "add" patches that extend the array, we can use a single "replace" patch (on the
// `length` property) as the inverse patch, instead of multiple "remove" patches.
```